pkg/origin-common/util/util.go

```diff
@@ -0,0 +1,300 @@
+		image.DockerImageMetadata.Comment = config.Comment
+		image.DockerImageMetadata.Created = config.Created
+		image.DockerImageMetadata.Container = config.Container
+		image.DockerImageMetadata.ContainerConfig = config.ContainerConfig
+		image.DockerImageMetadata.DockerVersion = config.DockerVersion
+		image.DockerImageMetadata.Author = config.Author
+		image.DockerImageMetadata.Config = config.Config
+		image.DockerImageMetadata.Architecture = config.Architecture
+	default:
+		return fmt.Errorf("unrecognized Docker image manifest schema %d for %q (%s)", manifest.SchemaVersion, image.Name, image.DockerImageReference)
+	}
+
+	layerSet := sets.NewString()
+	if manifest.SchemaVersion == 2 {
+		layerSet.Insert(manifest.Config.Digest)
+		image.DockerImageMetadata.Size = int64(len(image.DockerImageConfig))
+	} else {
+		image.DockerImageMetadata.Size = 0
+	}
+	for _, layer := range image.DockerImageLayers {
+		if layerSet.Has(layer.Name) {
+			continue
+		}
+		layerSet.Insert(layer.Name)
+		image.DockerImageMetadata.Size += layer.LayerSize
+	}
+
+	return nil
+}
+
+// ReorderImageLayers mutates the given image. It reorders the layers in ascending order.
+// Ascending order matches the order of layers in schema 2. Schema 1 has reversed (descending) order of layers.
+func ReorderImageLayers(image *imageapi.Image) {
+	if len(image.DockerImageLayers) == 0 {
+		return
+	}
+
+	layersOrder, ok := image.Annotations[imageapi.DockerImageLayersOrderAnnotation]
+	if !ok {
+		switch image.DockerImageManifestMediaType {
+		case schema1.MediaTypeManifest, schema1.MediaTypeSignedManifest:
+			layersOrder = imageapi.DockerImageLayersOrderAscending
+		case schema2.MediaTypeManifest:
+			layersOrder = imageapi.DockerImageLayersOrderDescending
+		default:
+			return
+		}
+	}
+
+	if layersOrder == imageapi.DockerImageLayersOrderDescending {
+		// reverse order of the layers (lowest = 0, highest = i)
+		for i, j := 0, len(image.DockerImageLayers)-1; i < j; i, j = i+1, j-1 {
+			image.DockerImageLayers[i], image.DockerImageLayers[j] = image.DockerImageLayers[j], image.DockerImageLayers[i]
+		}
+	}
+
+	if image.Annotations == nil {
+		image.Annotations = map[string]string{}
+	}
+
+	image.Annotations[imageapi.DockerImageLayersOrderAnnotation] = imageapi.DockerImageLayersOrderAscending
+}
+
+// ImageWithMetadata mutates the given image. It parses raw DockerImageManifest data stored in the image and
+// fills its DockerImageMetadata and other fields.
+// Copied from v3.7 github.com/openshift/origin/pkg/image/apis/image/v1/helpers.go
+func ImageWithMetadata(image *imageapiv1.Image) error {
+	// Check if the metadata are already filled in for this image.
+	meta, hasMetadata := image.DockerImageMetadata.Object.(*dockerapiv10.DockerImage)
+	if hasMetadata && meta.Size > 0 {
+		return nil
+	}
+
+	version := image.DockerImageMetadataVersion
+	if len(version) == 0 {
+		version = "1.0"
+	}
+
+	obj := &dockerapiv10.DockerImage{}
+	if len(image.DockerImageMetadata.Raw) != 0 {
```

Comment:
interesting.  had to do this since we don't have schema converters to decode the DockerImageMetadata.Raw data.  I don't feel great about it but for now there is only one "version" (pre12 has the same fields as 1.0) so this should always decode cleanly.

Reply:
I'd expect something like this: https://github.com/dmage/image-registry/commit/52af6c7546b457c841a1409c1df89e8f3fdfede7

Reply:
That does look better, thanks!